### DISCUSSION TITLE
[help] Is there a mapping pattern that can filter results?

### DISCUSSION THREAD

**Discussion by billdenney** [ropensci/targets#1014] [Help] [answered]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I have a target plan where I am loading many data files and then processing them in a pipeline.  Generally, I:

1. map across filenames to load the files,
2. map across the raw loaded results to have a cleaned and processed version,
3. detect problem files from step 2, and
4. map across the cleaned and processed version for the good files to generate final results.

(In the real version, there are several iterations of step 4 processing the cleaned results in several ways, but that gives the idea.)

Some of the source files have aspects that are invalid, so at the end of step 2 above, I want to have an indicator "File123.csv has an issue of X" come out of the cleaning (step 3), and I'd like to keep going with the valid files (step 4).

What I'd like to do is something like the following (which doesn't work) where I can say "run `myfun_process` only on good data".  Is something like that directly possible with `targets`?  The best programming pattern that I can think of to accomplish it is to make an intermediate list of the results which only include the good results and then map across that.  But, I want to be sure that I'm not missing something.

```
library(targets)
library(tarchetypes)

list(
  tar_target(
    filenames,
    list.files("source_data", full.names = TRUE)
  ),
  
  tar_target(
    d_raw,
    read.csv(file = filenames),
    pattern = map(filenames)
  ),
  
  tar_target(
    problems,
    myfun_detect_problems(d_raw),
    pattern = map(d_raw)
  ),
  
  tar_target(
    d_result,
    myfun_process(d_raw),
    pattern = map(d_raw[!problems])
  )
)
```

**Comment by wlandau** [comment, marked as answer]
There are built-in features that do vaguely similar thing, but I would recommend dropping the `problems` target and calling  `myfun_detect_problems()` inside `myfun_process()`. The latter can return `NULL` if `myfun_detect_problems()` finds a problem. Then `tar_read(d_result)` will aggregate the successful results with the `NULL`s. In the case of data frames, those `NULL`s will be dropped.

**Reply by billdenney** [reply]
Thanks.  That makes sense to me.  I currently do something similar, and I wanted to make sure that I wasn't missing a better method.

**Comment by stuvet** [comment]
Another (far less good, much more complicated, but very flexible) option I occasionally use is to run a statically branched pipeline fuelled by e.g. creating a vector of input filenames prepipeline, then censoring the same vector using a cached list of problem filenames, before running the pipeline as normal (defined by the filtered vector). Close the loop by appending problem filenames to the cache or a user-implemented state database from within a target.

It's far less elegant than the official solution, awkward to set up, probably awkward to maintain & certainly goes against some of what I'd suspect the `targets` ethos, but given appropriate static target names it allows some useful things like avoiding retrying (or even calculating whether or not to skip) a problem target until the next calendar day, exponential backoff for problem targets, staying within API limits, & dynamically prioritizing similar targets based on available API tokens. I recently used this to collect historic weather data, first collecting the most widely separated locations then filling in the blanks in between as daily API limits allowed.

I'd go with @wlandau unless you enjoy a challenge!
